src/Timer/Timers.php

```diff
@@ -77,32 +66,34 @@ public function isEmpty()
 
     public function tick()
     {
-        $time = $this->updateTime();
-        $timers = $this->timers;
-        $scheduler = $this->scheduler;
-
-        while (!$scheduler->isEmpty()) {
-            $timer = $scheduler->top();
+        // ensure timers are sorted so we can execute in order
+        if (!$this->sorted) {
+            $this->sorted = true;
+            asort($this->schedule);
+        }
 
-            if (!isset($timers[$timer])) {
-                $scheduler->extract();
-                $timers->detach($timer);
+        $time = $this->updateTime();
+        $schedule = $this->schedule;
 
-                continue;
+        foreach ($schedule as $id => $scheduled) {
+            // schedule is ordered, so loop until first timer that is not scheduled for execution now
+            if ($scheduled >= $time) {
+                break;
             }
 
-            if ($timers[$timer] >= $time) {
-                break;
+            // skip any timers that are removed while we process the current schedule
+            if (!isset($this->schedule[$id]) || $this->schedule[$id] !== $scheduled) {
+                continue;
             }
 
-            $scheduler->extract();
+            $timer = $this->timers[$id];
```

Comment:
Is it possible that this line will throw a notice of undefined index?

Reply:
I can see where you're coming from. This is not possible as both arrays are always modified pair-wise and as such the element is known to exist due to the check immediately before this assignment. I've added a comment to the code block below as this array may change *after* the timer has been invoked. Both cases are covered by the test suite, this code has 100% coverage.